[@property] Rules in shadow trees should be ignored.

anttijk opened the pull request:
#### 997d987fd5b6f1909b9fad9c0018f383a97ed8cc
<pre>
[@property] Rules in shadow trees should be ignored.
<a href="https://bugs.webkit.org/show_bug.cgi?id=250567">https://bugs.webkit.org/show_bug.cgi?id=250567</a>
rdar://104221943

Reviewed by Simon Fraser.

&quot;A @property is invalid if it occurs in a stylesheet inside of a shadow tree, and must be ignored.&quot;

<a href="https://drafts.css-houdini.org/css-properties-values-api/#at-property-rule">https://drafts.css-houdini.org/css-properties-values-api/#at-property-rule</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-shadow-expected.txt:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
(WebCore::DOMCSSRegisterCustomProperty::registerProperty):

No need to create a style resolver.

* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):

Fix indentation.

(WebCore::Style::RuleSetBuilder::addMutatingRulesToResolver):

The actual fix. Check if we are in a shadow scope and bail.

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::create):

Pass the scope as enum value (document or shadow). Due to sharing between identically styled
shadow trees we don&apos;t provide the actual Style::Scope.

(WebCore::Style::Resolver::Resolver):

Use WeakPtr for the document.

(WebCore::Style::Resolver::initialize):

Factor into a function.

(WebCore::Style::Resolver::document):
(WebCore::Style::Resolver::document const):
(WebCore::Style::Resolver::scope): Deleted.
(WebCore::Style::Resolver::scope const): Deleted.
(WebCore::Style::Resolver::~Resolver): Deleted.
* Source/WebCore/style/StyleResolver.h:
(WebCore::Style::Resolver::scopeType const):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::createDocumentResolver):
(WebCore::Style::Scope::createOrFindSharedShadowTreeResolver):

Canonical link: <a href="https://commits.webkit.org/258880@main">https://commits.webkit.org/258880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69078b64d50563644ba8061c4b6dc60377ced6c4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103251 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12376 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/36237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/112499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/172701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107204 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13402 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3280 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/95481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/110748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109025 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/36237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/95481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/92114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/36237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/95481 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5776 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/36237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5950 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/11936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/36237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6106 "Build is being retried. Recent messages:Cleaned up git repository; Checked out pull request; Verified commit is squashed; Reviewed by Simon Fraser; Validated commit message; Compiled WebKit (warnings); 14 flakes; Canonicalized commit; Pushed to pull request branch; 'git log ...'; Pushed commit to WebKit repository; Added comment on PR 8617 and added comment on bug 250567; Removed labels from pull request; Closed bug 250567") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7698 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
```diff
@@ -54,20 +54,16 @@ ExceptionOr<void> DOMCSSRegisterCustomProperty::registerProperty(Document& docum
-        auto parentStyle = RenderStyle::clone(*style);
-        Style::Builder dummyBuilder(*style, { document, parentStyle }, { }, { });
+        // We don't need to provide a real context style since only computationally independent values are allowed (no 'em' etc).
+        auto dummyStyle = RenderStyle::create();
```
I would prefer to avoid the word "dummy". Use "placeholder" or something instead.

smfr commented on Source/WebCore/style/StyleResolver.h:
```diff
@@ -151,14 +153,18 @@ class Resolver : public RefCounted<Resolver> {
     void applyMatchedProperties(State&, const MatchResult&);
 
+    WeakPtr<Document, WeakPtrImplWithEventTargetData> m_document;
+    ScopeType m_scopeType;
```
const

anttijk committed:
[@property] Rules in shadow trees should be ignored.
https://bugs.webkit.org/show_bug.cgi?id=250567
rdar://104221943

Reviewed by Simon Fraser.

"A @property is invalid if it occurs in a stylesheet inside of a shadow tree, and must be ignored."

https://drafts.css-houdini.org/css-properties-values-api/#at-property-rule

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-shadow-expected.txt:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
(WebCore::DOMCSSRegisterCustomProperty::registerProperty):

No need to create a style resolver.

* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):

Fix indentation.

(WebCore::Style::RuleSetBuilder::addMutatingRulesToResolver):

The actual fix. Check if we are in a shadow scope and bail.

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::create):

Pass the scope as enum value (document or shadow). Due to sharing between identically styled
shadow trees we don't provide the actual Style::Scope.

(WebCore::Style::Resolver::Resolver):

Use WeakPtr for the document.

(WebCore::Style::Resolver::initialize):

Factor into a function.

(WebCore::Style::Resolver::document):
(WebCore::Style::Resolver::document const):
(WebCore::Style::Resolver::scope): Deleted.
(WebCore::Style::Resolver::scope const): Deleted.
(WebCore::Style::Resolver::~Resolver): Deleted.
* Source/WebCore/style/StyleResolver.h:
(WebCore::Style::Resolver::scopeType const):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::createDocumentResolver):
(WebCore::Style::Scope::createOrFindSharedShadowTreeResolver):

Canonical link: https://commits.webkit.org/258880@main
